File: python/1218. Longest Arithmetic Subsequence of Given Difference.py
class Solution(object):
    def longestSubsequence(self, arr, difference):
        """
        :type arr: List[int]
        :type difference: int
        :rtype: int
        解法1: dfs超时
        """
        def dfs(idx, a):
            if idx==len(arr):
                return 0
            if a and arr[idx] - a[-1] != difference:
                return dfs(idx+1, a)
            return max(1 + dfs(idx+1, a+[arr[idx]]), dfs(idx+1, a))
        
        return dfs(0, [])

    
            

File: go/src/leetcode/1218. Longest Arithmetic Subsequence of Given Difference/1218. Longest Arithmetic Subsequence of Given Difference.go
/**
@author ZhengHao Lou
Date    2021/11/05
*/
package main

import "fmt"

/**
https://leetcode-cn.com/problems/longest-arithmetic-subsequence-of-given-difference/
 */
func longestSubsequence(arr []int, difference int) int {
	var ans int

	mapper := make(map[int]int)
	for _, num := range arr {
		if j, ok := mapper[num - difference]; ok {
			mapper[num] = j + 1
		} else {
			mapper[num] = 1
		}
		ans = max(ans, mapper[num])
	}
	fmt.Println(ans)
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	longestSubsequence([]int{1,2,3,4}, 1)
	longestSubsequence([]int{1,3,5,7}, 1)
	longestSubsequence([]int{1,5,7,8,5,3,4,2,1}, -2)
	longestSubsequence([]int{3,0,-3,4,-4,7,6}, 3) // 2
}
